File: sdk/python/agentfield/harness/providers/claude.py
"""Claude Code provider using claude_agent_sdk (native Python SDK).

Uses lazy import - claude_agent_sdk is an optional dependency that's only
loaded when the claude-code provider is actually used.
"""

from __future__ import annotations

import time
from typing import Any, Dict, List, Optional

from agentfield.harness._result import Metrics, RawResult


def _get_claude_sdk() -> Any:
    """Lazy import of claude_agent_sdk."""
    try:
        import claude_agent_sdk  # pyright: ignore[reportMissingImports]

        return claude_agent_sdk
    except ImportError as exc:
        raise ImportError(
            "claude_agent_sdk is required for the 'claude-code' provider. "
            "Install it with: pip install claude-agent-sdk"
        ) from exc


_PERMISSION_MAP = {
    "auto": "bypassPermissions",
    "plan": "plan",
}


class ClaudeCodeProvider:
    """Claude Code provider using the native claude_agent_sdk."""

    async def execute(self, prompt: str, options: dict[str, object]) -> RawResult:
        """Execute a prompt via Claude Code SDK."""
        sdk = _get_claude_sdk()

        agent_options: dict[str, object] = {}
        if options.get("model") is not None:
            agent_options["model"] = options["model"]
        if options.get("cwd") is not None:
            agent_options["cwd"] = options["cwd"]
        if options.get("max_turns") is not None:
            agent_options["max_turns"] = options["max_turns"]
        if options.get("tools") is not None:
            agent_options["allowed_tools"] = options["tools"]
        if options.get("system_prompt") is not None:
            agent_options["system_prompt"] = options["system_prompt"]
        if options.get("max_budget_usd") is not None:
            agent_options["max_budget_usd"] = options["max_budget_usd"]
        if options.get("permission_mode") is not None:
            raw_mode = str(options["permission_mode"])
            agent_options["permission_mode"] = _PERMISSION_MAP.get(raw_mode, raw_mode)
        if options.get("env") is not None:
            agent_options["env"] = options["env"]

        messages: List[Dict[str, Any]] = []
        result_text: Optional[str] = None
        total_cost: Optional[float] = None
        num_turns = 0
        session_id = ""
        start_api = time.monotonic()

        try:
            opts = (
                sdk.ClaudeAgentOptions(**agent_options)
                if hasattr(sdk, "ClaudeAgentOptions")
                else agent_options
            )

            async for msg in sdk.query(prompt=prompt, options=opts):
                if isinstance(msg, dict):
                    msg_dict = msg
                elif hasattr(msg, "__dict__"):
                    msg_dict = dict(msg.__dict__)
                else:
                    msg_dict = {"raw": str(msg)}

                messages.append(msg_dict)

                msg_type = str(msg_dict.get("type", ""))
                if msg_type == "result":
                    raw_result = msg_dict.get("result", msg_dict.get("text", ""))
                    result_text = (
                        raw_result if isinstance(raw_result, str) else str(raw_result)
                    )
                    sid = msg_dict.get("session_id", "")
                    session_id = sid if isinstance(sid, str) else str(sid)
                    cost_info = msg_dict.get("cost_usd") or msg_dict.get(
                        "total_cost_usd"
                    )
                    if cost_info is not None:
                        total_cost = float(cost_info)
                    turns = msg_dict.get("num_turns")
                    num_turns = (
                        int(turns) if isinstance(turns, (int, float)) else len(messages)
                    )
                elif msg_type == "assistant" and result_text is None:
                    content = msg_dict.get("content")
                    message_obj = msg_dict.get("message")
                    if content is None and isinstance(message_obj, dict):
                        content = message_obj.get("content")

                    if isinstance(content, str):
                        result_text = content
                    elif isinstance(content, list):
                        for block in content:
                            if isinstance(block, dict) and block.get("type") == "text":
                                text = block.get("text")
                                if isinstance(text, str):
                                    result_text = text

            api_ms = int((time.monotonic() - start_api) * 1000)

            return RawResult(
                result=result_text,
                messages=messages,
                metrics=Metrics(
                    duration_ms=0,
                    duration_api_ms=api_ms,
                    num_turns=num_turns,
                    total_cost_usd=total_cost,
                    session_id=session_id,
                ),
                is_error=False,
            )
        except Exception as exc:
            api_ms = int((time.monotonic() - start_api) * 1000)
            return RawResult(
                result=None,
                messages=messages,
                metrics=Metrics(duration_api_ms=api_ms, session_id=session_id),
                is_error=True,
                error_message=str(exc),
            )

File: sdk/typescript/src/harness/providers/claude.ts
import type { HarnessProvider } from './base.js';
import type { RawResult } from '../types.js';
import { createMetrics, createRawResult } from '../types.js';

type QueryInput = {
  prompt: string;
  options: Record<string, unknown>;
};

type ClaudeSdkModule = {
  query: (input: QueryInput) => AsyncIterable<unknown>;
};

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null;
}

function getString(record: Record<string, unknown>, key: string): string | undefined {
  const value = record[key];
  return typeof value === 'string' ? value : undefined;
}

function getNumber(record: Record<string, unknown>, key: string): number | undefined {
  const value = record[key];
  return typeof value === 'number' ? value : undefined;
}

export class ClaudeCodeProvider implements HarnessProvider {
  public async execute(prompt: string, options: Record<string, unknown>): Promise<RawResult> {
    let sdk: ClaudeSdkModule;
    try {
      const mod = await import('@anthropic-ai/claude-agent-sdk');
      sdk = mod as ClaudeSdkModule;
    } catch {
      throw new Error(
        "@anthropic-ai/claude-agent-sdk is required for the 'claude-code' provider. " +
          'Install it with: npm install @anthropic-ai/claude-agent-sdk'
      );
    }

    const agentOptions: Record<string, unknown> = {};
    if (options.model !== undefined) agentOptions.model = options.model;
    if (options.cwd !== undefined) agentOptions.cwd = options.cwd;
    if (options.maxTurns !== undefined) agentOptions.max_turns = options.maxTurns;
    if (options.tools !== undefined) agentOptions.allowed_tools = options.tools;
    if (options.systemPrompt !== undefined) agentOptions.system_prompt = options.systemPrompt;
    if (options.maxBudgetUsd !== undefined) agentOptions.max_budget_usd = options.maxBudgetUsd;
    if (options.permissionMode !== undefined) {
      const modeMap: Record<string, string> = { auto: 'bypassPermissions', plan: 'plan' };
      const raw = String(options.permissionMode);
      agentOptions.permission_mode = modeMap[raw] ?? raw;
    }
    if (options.env !== undefined) agentOptions.env = options.env;

    const messages: Array<Record<string, unknown>> = [];
    let resultText: string | undefined;
    let totalCost: number | undefined;
    let numTurns = 0;
    let sessionId = '';
    const startApi = Date.now();

    try {
      for await (const msg of sdk.query({ prompt, options: agentOptions })) {
        const msgObj = isRecord(msg) ? msg : { raw: String(msg) };
        messages.push(msgObj);

        const msgType = getString(msgObj, 'type') ?? '';
        if (msgType === 'result') {
          const resultValue = msgObj.result ?? msgObj.text;
          resultText = typeof resultValue === 'string' ? resultValue : resultValue == null ? '' : String(resultValue);

          const sid = getString(msgObj, 'session_id');
          if (sid !== undefined) {
            sessionId = sid;
          }

          const costUsd = getNumber(msgObj, 'cost_usd');
          const totalCostUsd = getNumber(msgObj, 'total_cost_usd');
          if (costUsd !== undefined) {
            totalCost = costUsd;
          } else if (totalCostUsd !== undefined) {
            totalCost = totalCostUsd;
          }

          const turns = getNumber(msgObj, 'num_turns');
          numTurns = turns === undefined ? messages.length : Math.trunc(turns);
        } else if (msgType === 'assistant' && resultText === undefined) {
          let content: unknown = msgObj.content;
          if (content === undefined && isRecord(msgObj.message)) {
            content = msgObj.message.content;
          }

          if (typeof content === 'string') {
            resultText = content;
          } else if (Array.isArray(content)) {
            for (const block of content) {
              if (isRecord(block) && block.type === 'text' && typeof block.text === 'string') {
                resultText = block.text;
              }
            }
          }
        }
      }

      return createRawResult({
        result: resultText,
        messages,
        metrics: createMetrics({
          durationApiMs: Date.now() - startApi,
          numTurns,
          totalCostUsd: totalCost,
          sessionId,
        }),
        isError: false,
      });
    } catch (error: unknown) {
      return createRawResult({
        result: undefined,
        messages,
        metrics: createMetrics({ durationApiMs: Date.now() - startApi, sessionId }),
        isError: true,
        errorMessage: error instanceof Error ? error.message : String(error),
      });
    }
  }
}
